js/modules/k6/ws/ws.go

```diff
@@ -0,0 +1,385 @@
+	netDial := func(network, address string) (net.Conn, error) {
+		return state.Dialer.DialContext(ctx, network, address)
+	}
+
+	wsd := websocket.Dialer{
+		NetDial: netDial,
+		Proxy:   http.ProxyFromEnvironment,
+	}
+
+	start := time.Now()
+	conn, response, connErr := wsd.Dial(url, header)
+	connectionEnd := time.Now()
+	connectionDuration := stats.D(connectionEnd.Sub(start))
+
+	socket := Socket{
+		ctx:                ctx,
+		conn:               conn,
+		eventHandlers:      make(map[string][]goja.Callable),
+		pingSendTimestamps: make(map[string]time.Time),
+		scheduled:          make(chan goja.Callable),
+		done:               make(chan struct{}),
+	}
+
+	// Run the user-provided set up function
+	setupFn(goja.Undefined(), rt.ToValue(&socket))
+
+	if connErr != nil {
+		// Pass the error to the user script before exiting immediately
+		socket.handleEvent("error", rt.ToValue(connErr))
+		return response
+	}
+	defer conn.Close()
+
+	tags["status"] = strconv.Itoa(response.StatusCode)
+	tags["subprotocol"] = response.Header.Get("Sec-WebSocket-Protocol")
+
+	// The connection is now open, emit the event
+	socket.handleEvent("open")
+
+	// Pass ping/pong events through the main control loop
+	pingChan := make(chan string)
+	pongChan := make(chan string)
+	conn.SetPingHandler(func(msg string) error { pingChan <- msg; return nil })
+	conn.SetPongHandler(func(pingID string) error { pongChan <- pingID; return nil })
+
+	readDataChan := make(chan []byte)
+	readErrChan := make(chan error)
+
+	// Wraps a couple of channels around conn.ReadMessage
+	go readPump(conn, readDataChan, readErrChan)
+
+	// This is the main control loop. All JS code (including error handlers)
+	// should only be executed by this thread to avoid race conditions
+	for {
+		select {
+		case <-pingChan:
+			// Handle pings received from the server
+			socket.handleEvent("ping")
+
+		case pingID := <-pongChan:
+			// Handle pong responses to our pings
+			socket.trackPong(pingID)
+			socket.handleEvent("pong")
+
+		case readData := <-readDataChan:
+			socket.msgsReceived++
+			socket.handleEvent("message", rt.ToValue(string(readData)))
+
+		case readErr := <-readErrChan:
+			socket.handleEvent("error", rt.ToValue(readErr))
+
+		case scheduledFn := <-socket.scheduled:
+			scheduledFn(goja.Undefined())
+
+		case <-ctx.Done():
+			// This means that K6 is shutting down (e.g., during an interrupt)
+			socket.handleEvent("close", rt.ToValue("Interrupt"))
+			socket.closeConnection(websocket.CloseGoingAway)
+
+		case <-socket.done:
+			// This is the final exit point normally triggered by closeConnection
+			end := time.Now()
+			sessionDuration := stats.D(end.Sub(start))
+
+			samples := []stats.Sample{
+				{Metric: metrics.WSSessions, Time: end, Tags: tags, Value: 1},
+				{Metric: metrics.WSMessagesSent, Time: end, Tags: tags, Value: socket.msgsReceived},
+				{Metric: metrics.WSMessagesReceived, Time: end, Tags: tags, Value: socket.msgsSent},
```

Comment:
Can you take these sent/received counters the way of the ping counter and have one emission per message instead? A series of 1's make a much more telling graph than a huge spike like this.